File: src/Atlasd/Battlenet/Protocols/Game/ChatCommands/WhisperCommand.cs
using Atlasd.Localization;
using System.Collections.Generic;
using System.Text;

namespace Atlasd.Battlenet.Protocols.Game.ChatCommands
{
    class WhisperCommand : ChatCommand
    {
        public WhisperCommand(byte[] rawBuffer, List<string> arguments) : base(rawBuffer, arguments) { }

        public override bool CanInvoke(ChatCommandContext context)
        {
            return context != null && context.GameState != null && context.GameState.ActiveAccount != null;
        }

        public override void Invoke(ChatCommandContext context)
        {
            if (Arguments.Count < 1)
            {
                new ChatEvent(ChatEvent.EventIds.EID_ERROR, context.GameState.ChannelFlags, context.GameState.Ping, context.GameState.OnlineName, Resources.UserNotLoggedOn).WriteTo(context.GameState.Client);
                return;
            }

            var target = Arguments[0];
            Arguments.RemoveAt(0);
            // Calculates and removes (target+' ') from (raw) which prints into (newRaw):
            RawBuffer = RawBuffer[(Encoding.UTF8.GetByteCount(target) + (Arguments.Count > 0 ? 1 : 0))..];

            // Check for empty message
            if (RawBuffer.Length == 0)
            {
                new ChatEvent(ChatEvent.EventIds.EID_ERROR, context.GameState.ChannelFlags, context.GameState.Ping, context.GameState.OnlineName, Resources.WhisperCommandEmptyMessage).WriteTo(context.GameState.Client);
                return;
            }

            // Get the target state, or return not logged on
            if (!Battlenet.Common.GetClientByOnlineName(target, out var targetState) || targetState == null)
            {
                new ChatEvent(ChatEvent.EventIds.EID_ERROR, context.GameState.ChannelFlags, context.GameState.Ping, context.GameState.OnlineName, Resources.UserNotLoggedOn).WriteTo(context.GameState.Client);
                return;
            }

            // Check if target is asking to not be disturbed
            if (!string.IsNullOrEmpty(targetState.DoNotDisturb))
            {
                var r = Resources.WhisperCommandUserIsDoNotDisturb;

                r = r.Replace("{user}", targetState.OnlineName);
                r = r.Replace("{message}", targetState.Away);

                new ChatEvent(ChatEvent.EventIds.EID_INFO, context.GameState.ChannelFlags, context.GameState.Ping, context.GameState.OnlineName, r).WriteTo(context.GameState.Client);

                return; // Target has asked to not be disturbed, discontinue
            }

            // Notify the source we whispered the target
            new ChatEvent(ChatEvent.EventIds.EID_WHISPERTO, Channel.RenderChannelFlags(context.GameState, targetState), targetState.Ping, Channel.RenderOnlineName(context.GameState, targetState), RawBuffer).WriteTo(context.GameState.Client);

            // Check if target is marked away
            if (!string.IsNullOrEmpty(targetState.Away))
            {
                var r = Resources.WhisperCommandUserIsAway;

                r = r.Replace("{user}", targetState.OnlineName);
                r = r.Replace("{message}", targetState.Away);

                new ChatEvent(ChatEvent.EventIds.EID_INFO, Channel.RenderChannelFlags(context.GameState, targetState), context.GameState.Ping, Channel.RenderOnlineName(context.GameState, targetState), r).WriteTo(context.GameState.Client);
            }

            // Notify the target they were whispered by the source
            new ChatEvent(ChatEvent.EventIds.EID_WHISPERFROM, Channel.RenderChannelFlags(targetState, context.GameState), context.GameState.Ping, Channel.RenderOnlineName(targetState, context.GameState), RawBuffer).WriteTo(targetState.Client);
        }
    }
}

File: src/AtlasV/Battlenet/Protocols/Game/ChatCommands/WhisperCommand.vb
Imports AtlasV.Localization
Imports System.Collections.Generic
Imports System.Net
Imports System.Text

Namespace AtlasV.Battlenet.Protocols.Game.ChatCommands
    Class WhisperCommand
        Inherits ChatCommand

        Public Sub New(ByVal varRawBuffer As Byte(), ByVal varArguments As List(Of String))
            MyBase.New(varRawBuffer, varArguments)
        End Sub

        Public Overrides Function CanInvoke(ByVal varContext As ChatCommandContext) As Boolean
            Return varContext IsNot Nothing AndAlso varContext.GameState IsNot Nothing AndAlso varContext.GameState.ActiveAccount IsNot Nothing
        End Function

        Public Overrides Sub Invoke(ByVal varContext As ChatCommandContext)
            If Arguments.Count < 1 Then
                Call New ChatEvent(ChatEvent.EventIds.EID_ERROR, varContext.GameState.ChannelFlags, varContext.GameState.Ping, varContext.GameState.OnlineName, Resources.UserNotLoggedOn).WriteTo(varContext.GameState.Client)
                Return
            End If

            Dim target = Arguments(0)
            Arguments.RemoveAt(0)
            RawBuffer = RawBuffer.Skip((Encoding.UTF8.GetByteCount(target) + (If(Arguments.Count > 0, 1, 0)))).ToArray()

            If RawBuffer.Length = 0 Then
                Call New ChatEvent(ChatEvent.EventIds.EID_ERROR, varContext.GameState.ChannelFlags, varContext.GameState.Ping, varContext.GameState.OnlineName, Resources.WhisperCommandEmptyMessage).WriteTo(varContext.GameState.Client)
                Return
            End If

            Dim targetState As GameState = Nothing

            If Not Battlenet.Common.GetClientByOnlineName(target, targetState) OrElse targetState Is Nothing Then
                Call New ChatEvent(ChatEvent.EventIds.EID_ERROR, varContext.GameState.ChannelFlags, varContext.GameState.Ping, varContext.GameState.OnlineName, Resources.UserNotLoggedOn).WriteTo(varContext.GameState.Client)
                Return
            End If

            If Not String.IsNullOrEmpty(targetState.DoNotDisturb) Then
                Dim r = Resources.WhisperCommandUserIsDoNotDisturb
                r = r.Replace("{user}", targetState.OnlineName)
                r = r.Replace("{message}", targetState.Away)
                Call New ChatEvent(ChatEvent.EventIds.EID_INFO, varContext.GameState.ChannelFlags, varContext.GameState.Ping, varContext.GameState.OnlineName, r).WriteTo(varContext.GameState.Client)
                Return
            End If

            Call New ChatEvent(ChatEvent.EventIds.EID_WHISPERTO, Channel.RenderChannelFlags(varContext.GameState, targetState), targetState.Ping, Channel.RenderOnlineName(varContext.GameState, targetState), RawBuffer).WriteTo(varContext.GameState.Client)

            If Not String.IsNullOrEmpty(targetState.Away) Then
                Dim r = Resources.WhisperCommandUserIsAway
                r = r.Replace("{user}", targetState.OnlineName)
                r = r.Replace("{message}", targetState.Away)
                Call New ChatEvent(ChatEvent.EventIds.EID_INFO, Channel.RenderChannelFlags(varContext.GameState, targetState), varContext.GameState.Ping, Channel.RenderOnlineName(varContext.GameState, targetState), r).WriteTo(varContext.GameState.Client)
            End If

            If targetState.SquelchedIPs.Contains(IPAddress.Parse(varContext.GameState.Client.RemoteEndPoint.ToString().Split(":"c)(0))) Then
                Return
            End If

            Call New ChatEvent(ChatEvent.EventIds.EID_WHISPERFROM, Channel.RenderChannelFlags(targetState, varContext.GameState), varContext.GameState.Ping, Channel.RenderOnlineName(targetState, varContext.GameState), RawBuffer).WriteTo(targetState.Client)
        End Sub
    End Class
End Namespace
